### DISCUSSION TITLE
Error in using schema.rs from models.rs

### DISCUSSION THREAD

**Discussion by friendbear** [diesel-rs/diesel#2979] [Q&A] [answered]
The environment is as follows, and the official diesel tutorial 1 worked fine.

````
rustc --version && cargo --version
rustc 1.57.0 (f1edd0429 2021-11-29)
cargo 1.57.0 (b2e52d7ca 2021-10-21)
````

However, [a simple Web Server](https://gill.net.in/posts/auth-microservice-rust-actix-web1.0-diesel-complete-tutorial/) with two tables , I just can't get rid of the notation problem.

first 

```models.rs
use diesel::{r2d2::ConnectionManager, PgConnection};
use serde::{Deserialize, Serialize};
use super::schema::*;

// or use super::schema::{users, invitations};
```

1. cargo check ...

> src/models.rs:6:12
  |... > src/models.rs:6:12
6 | use super::schema::{users, invitations}
  | ^^^^^^
  | unresolved import
  | unresolved import
  | help: a similar path exists: `auth_service::schema`.

However, if you specify the use clause for only one table, the error for that table will be removed.
```
use super::schema::users;
```

I'm very sorry, but could you please help me?

**Comment by weiznich** [comment, marked as answer]
This seems to be an issue with the way how you setup your module structure. It seems like you have a separate binary/library crate structure, but it's hard to tell that exactly without knowing details. Can you provide a full example how to reproduce this?

**Reply by friendbear** [reply]
@weiznich san.

Thank you for your reply.
The code in question is being implemented here.

<https://github.com/friendbear/auth-service/tree/diesel-discussions-2979>

It is only implemented up to the content of the question, and it does not work, but I hope you can check it.

**Reply by weiznich** [reply]
Ok the error mentioned in your first post happens because you include the models module from both the [`main.rs`](https://github.com/friendbear/auth-service/blob/diesel-discussions-2979/src/main.rs) and the [`lib.rs`](https://github.com/friendbear/auth-service/blob/diesel-discussions-2979/src/lib.rs#L2) file. Your `lib.rs` file also contains a `schema` module, while your `main.rs` file does not. You can thing about `lib.rs` vs `main.rs` as being two distinct crates. This implies there is just no `schema` module for the crate created by the `main.rs` file. Generally speaking if you want to have both a library and a binary you want to put everything that should be reusable as library into the `lib.rs` file and then import that crate via `use your_crate_name::some_module` in your `main.rs`. It's an antipattern to include modules in both "crates" via the `mod` keyword.

**Reply by friendbear** [reply]
Thank you very much for your very clear and precise answer.
I'm sorry that I don't understand the anti-pattern.

The errors in models.rs disappeared when I removed models from main.rs.
I will review the structure.

Thank you very much. 💚❄
